feat: add support for csv files (#592)
jina-ai/finetuner: docs/walkthrough/create-training-data.md

(create-training-data)=
# Prepare Training Data

Finetuner accepts training data and evaluation data in the form of csv files or{class}`~docarray.array.document.DocumentArray` objects.
Because Finetuner follows a [supervised-learning](https://en.wikipedia.org/wiki/Supervised_learning) scheme, each element requires a label that identifies which other elements it should be similar to. 
If you need to evaluate metrics on separate evaluation data, it is recommended to dataset only for evaluation purposes.

> Data can be prepared in two different forms, either as a csv file, or a {class}`~docarray.array.document.DocumentArray`. In the sections below, you can see examples which demonstrate how the training datasets should look like for each form: 

Reviewer: ```suggestion
Data can be prepared in two different formats, either as a CSV file, or as a {class}`~docarray.array.document.DocumentArray`. In the sections below, you can see examples which demonstrate how the training datasets should look like for each format.
```

> Data can be prepared in two different forms, either as a csv file, or a {class}`~docarray.array.document.DocumentArray`. In the sections below, you can see examples which demonstrate how the training datasets should look like for each form: 

Reviewer: ```suggestion
Data can be prepared in two different formats, either as a csv file, or a {class}`~docarray.array.document.DocumentArray`. In the sections below, you can see examples which demonstrate how the training datasets should look like for each form: 
```

## Preparing CSV Files

> To record data in a csv file, the contents of each element are stored plainly, with each row either representing one labelled item, multiple items that should be semantically similar, or two items of different modalities in the case that a CLIP model is being used.

Reviewer: ```suggestion
To record data in a CSV file, the contents of each element are stored plainly, with each row either representing one labeled item, multiple items that should be semantically similar, or two items of different modalities in the case that a CLIP model is being used.
```

> To record data in a csv file, the contents of each element are stored plainly, with each row either representing one labelled item, multiple items that should be semantically similar, or two items of different modalities in the case that a CLIP model is being used.

Reviewer: multiple or just 2?

Author: just 2, I will clarify

> Currently, `excel`, `excel-tab` and `unix` csv dialects are supported. To specify which dialect to use, provide a dictionary with `['dialect']=chosen_dialect` as the `csv_options` argument to the {meth}`~finetuner.fit` function. The list of all options for reading csv files can be found in the description of the {meth}`~finetuner.utils.from_csv` function.

Reviewer: ```suggestion
Currently, `excel`, `excel-tab` and `unix` CSV dialects are supported. To specify which dialect to use, provide a dictionary with `['dialect']=chosen_dialect` as the `csv_options` argument to the {meth}`~finetuner.fit` function. The list of all options for reading CSV files can be found in the description of the {meth}`~finetuner.utils.from_csv` function.
```

> ````{tab} multiple elements per row

Reviewer: ```suggestion
````{tab} Multiple elements per row
```

If you want two elements to be semantically close together, they can be placed on the same row as a pair:

```markdown
This is an English sentence         Das ist ein englischer Satz
This is another English sentence    Dies ist ein weiterer englischer Satz
...
```

> This format can be used for data for both text-to-text and image-to-image:

Reviewer: ```suggestion
This format can be used for data for both text-to-text and image-to-image tasks:
```

```markdown
apple.jpg   https://example.com/apple-styling.jpg
orange.jpg  https://example.com/orange-styling.jpg
```
````

> ````{tab} labeled data

Reviewer: ```suggestion
````{tab} Labeled data
```

In cases where you want multiple elements grouped together, you can provide a label in the second column. This way, all elements in the first column that have the same label will be considered similar when training. To indicate that the second column of your csv file represents a label instead of a second element, make sure to set `is_labeled = True` in the `csv_options` argument of the {meth}`~finetuner.fit` function. Your data can then be structured like so:

```markdown
Hello!                  greeting-english
Hi there.               greeting-english
Good morning.           greeting-english
I'm (…) sorry!          apologize-english
I'm sorry to have…      apologize-english
Please, forgive me!     apologize-english
```
````

> ````{tab} text-to-image search on CLIP

Reviewer: ```suggestion
````{tab} Data format for text-to-image search using CLIP
```

To prepare data for text-to-image search, each row must contain one uri to an image, and one piece of text. The order that these two are placed does not matter, so long as the ordering is kept consistent for all rows.

```markdown
This is a photo of an apple.                apple.jpg
This is a black-white photo of an organge.  orange.jpg
```

```{admonition} CLIP model explained
:class: hint
OpenAI CLIP model wraps two models: a vision transformer and a text transformer.
During fine-tuning, we're optimizing two models in parallel.

At the model saving time, you will discover, we are saving two models to your local directory. 
```

````
## Preparing a DocumentArray

> When providing training data in a DocumentArray, each element is represented as a {class}`~docarray.document.Document`. you should assign a label to each {class}`~docarray.document.Document` inside your {class}`~docarray.array.document.DocumentArray`.

Reviewer: ```suggestion
When providing training data in a DocumentArray, each element is represented as a {class}`~docarray.document.Document`. You should assign a label to each {class}`~docarray.document.Document` inside your {class}`~docarray.array.document.DocumentArray`.
```

```markdown
Hello!                  greeting-english
Hi there.               greeting-english
Good morning.           greeting-english
I'm (…) sorry!          apologize-english
I'm sorry to have…      apologize-english
Please, forgive me!     apologize-english
```
````

````{tab} text-to-image search on CLIP
To prepare data for text-to-image search, each row must contain one uri to an image, and one piece of text. The order that these two are placed does not matter, so long as the ordering is kept consistent for all rows.

```markdown
This is a photo of an apple.                apple.jpg
This is a black-white photo of an organge.  orange.jpg
```

```{admonition} CLIP model explained
:class: hint
OpenAI CLIP model wraps two models: a vision transformer and a text transformer.
During fine-tuning, we're optimizing two models in parallel.

At the model saving time, you will discover, we are saving two models to your local directory. 

> ```

Reviewer 2: now the same message duplicated twice, remove the one below
